src/spectrum_assignment.py

```diff
@@ -107,25 +109,91 @@ def handle_first_last(self, flag: str):
+                                for cnt_tmp2, slot in enumerate(list(row)):
+                                    was_allocated = self.spec_help_obj.check_super_channels(open_slots_matrix=open_slots_matrix, flag=flag)
+                                    snr_result = self.snr_obj.handle_snr_dynamic_slicing(self.sdn_props.path_index)[0]
+                                    if snr_result is None:
+                                        if row:
+                                            open_slots_matrix[cnt_tmp].pop(0)                                        
+                                        if all(len(row1) == 0 for row1 in open_slots_matrix):
+                                            was_allocated = False
+                                            break
+                                    else:
+                                        was_allocated = True
+                                        break_outer_loop = True
+                                        break
+                                if break_outer_loop: 
+                                    break
+                    if was_allocated:
+                        return
+        else:
+            for core_arr, core_num in zip(core_matrix, core_list):
+                for band_index in range(len(band_list)):  # pylint: disable=consider-using-enumerate
+                    open_slots_arr = np.where(core_arr[band_index] == 0)[0]
+
+                    # Source: https://stackoverflow.com/questions/3149440/splitting-list-based-on-missing-numbers-in-a-sequence
+                    if flag in ('last_fit', 'priority_last'):
+                        open_slots_matrix = [list(map(itemgetter(1), g))[::-1] for k, g in
+                                            itertools.groupby(enumerate(open_slots_arr), lambda i_x: i_x[0] - i_x[1])]
+                    elif flag in ('first_fit', 'priority_first', 'forced_index'):
+                        open_slots_matrix = [list(map(itemgetter(1), g)) for k, g in
+                                            itertools.groupby(enumerate(open_slots_arr), lambda i_x: i_x[0] - i_x[1])]
+                    else:
+                        raise NotImplementedError(f'Invalid flag, got: {flag} and expected last_fit or first_fit.')
+
+                    self.spec_help_obj.core_num = core_num
+                    self.spec_help_obj.curr_band = band_list[band_index]
+                    was_allocated = self.spec_help_obj.check_super_channels(open_slots_matrix=open_slots_matrix, flag=flag)
+                    # support FF for the spectrum meet the minimum gsnr requirments
+                    if self.engine_props['cores_per_link'] in [13, 19] and self.engine_props['snr_type'] == 'snr_e2e_external_resources':
+                        snr_result = self.snr_obj.handle_snr_dynamic_slicing(self.sdn_props.path_index)[0]
+                        break_outer_loop = False 
+
+                        if snr_result is None:
```

Comment:
This looks similar to the above code. We can probably add a helper function to interact with both parts with flags as needed.